zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/InboundService.kt

```diff
@@ -68,47 +68,88 @@ internal class InboundService<T : ZiplineService>(
     externalCall: Call,
     suspendCallback: SuspendCallback<Any?>,
   ): String {
-    val job = endpoint.scope.launch {
-      @Suppress("UNCHECKED_CAST") // We found this function by matching on its service type T.
-      val function = internalCall.function as SuspendingZiplineFunction<ZiplineService>
-      val args = internalCall.args
+    @Suppress("UNCHECKED_CAST") // We found this function by matching on its service type T.
+    val function = internalCall.function as SuspendingZiplineFunction<ZiplineService>
 
-      val callStart = endpoint.eventListener.callStart(externalCall)
-      val result = runCatching {
+    val callStart = endpoint.eventListener.callStart(externalCall)
+
+    val deferred = endpoint.scope.async(start = UNDISPATCHED) {
+      val args = internalCall.args
+      runCatching {
         function.callSuspending(service, args)
       }
-      endpoint.callCodec.nextOutboundCallCallback = { callbackCall ->
-        endpoint.eventListener.callEnd(
-          externalCall,
-          CallResult(result, callbackCall.encodedCall, callbackCall.serviceNames),
-          callStart,
-        )
-      }
+    }
 
-      // Don't resume a continuation if the Zipline has since been closed.
-      endpoint.scope.ensureActive()
+    if (deferred.isActive) {
+      val cancelCallback = object : CancelCallback, HasPassByReferenceName {
+        override var passbyReferenceName: String? = null
 
-      when {
-        result.isSuccess -> suspendCallback.success(result.getOrNull())
-        else -> suspendCallback.failure(result.exceptionOrNull()!!)
+        override fun cancel() {
+          deferred.cancel()
+        }
+
+        override fun toString() = "CancelCallback/$internalCall"
       }
-    }
 
-    val cancelCallback = object : CancelCallback, HasPassByReferenceName {
-      override var passbyReferenceName: String? = null
+      deferred.invokeOnCompletion {
```

Comment:
I'm a bit worried about the Retrofit case here: where `isActive` returns `true` but by the time you get to this code the `async` has completed and thus this lambda will be invoked synchronously (I'm assuming that's the behavior).

It would mean that we invoke the `suspendCallback` below synchronously from the stackframe which invoked the `suspend` function (and before we allow it to return a value). The normal coroutine machinery defends against this (which is what was happening in Retrofit).

In our case, to defend against it, I think we need something like the following:
```kotlin
var scheduleCallback = true
deferred.invokeOnCompletion {
  // ..

  if (scheduleCallback) {
    ziplineDispatcher.schedule({ /* callback logic */ })
  } else {
    // callback logic
  }
}
scheduleCallback = false

Reply:
It’s single-threaded, so we don‘t need to worry about that.

Reply:
As in, the `async` is running on the same dispatcher as this block? That's good.

Reply:
Exactly.